### ISSUE TITLE
Misleading memory requirements

### ISSUE THREAD

**Issue by letmaik** [lsds/sgx-lkl#208]
As a user, if I want to evaluate how much memory SGX-LKL will consume then I may look at the verbose output and discover this:
```
[[  SGX-LKL ]] sgxlkl_enclave_show_attribute(): enclave base=0x7fb900000000 size=1.042 GB
[[  SGX-LKL ]] sgxlkl_enclave_show_attribute(): enclave heap base=0x7fb900aa0000 size=1024.00 M end=0x7fb940aa0000
```
I may assume that the enclave itself needs 1.042 GB of memory. Accounting for a tiny amount of host-side memory gives a rough estimate of 1.05 GB total memory.

However, this is a lie. Here is what actually happens:
- When OE creates an enclave, it aligns 1.042 GB to the [next power of two](https://github.com/openenclave/openenclave/blob/91d3069bf072addeab6151d1cfe8a39d1dd529be/host/sgx/create.c#L313-L317), which becomes the actual hardware enclave size (2 GB). This is an SGX requirement. All the extra enclave space is essentially unused. In software mode, the actual [allocated memory for the enclave is doubled](https://github.com/openenclave/openenclave/blob/2490b0268f1dcf346dd977c81ac90ba73546ff9c/host/sgx/sgxload.c#L183-L196) on top of that, bringing it to 4 GB.
- OE then sets 1.042 GB as "visible" enclave size, which is what SGX-LKL is dumping out.

Note that the difference between the enclave size and heap size is roughly the SGX-LKL ELF image size plus some extra stack and control pages, in this case ~43 MB. This is an important number because it can be used to optimize memory usage by lowering heap size such that the enclave size stays below the next power of two, avoiding wasting memory (e.g. 975 MB would be a good number).

I think this issue can be solved with better tooling and documentation. SGX-LKL should also output a warning message if it detects that the heap size is chosen badly leading to >80% unused enclave memory.

**Comment by letmaik** [comment]
@jhand2 Can you check if what I said is correct? @prp has some doubts that the unused enclave memory is not actually counted towards process space.

**Comment by letmaik** [comment]
@jhand2 @jxyang Any insights?

**Comment by jhand2** [comment]
Honestly I'm not sure if the remaining unused size counts as taking up EPC (I would guess it does but I don't know). @CodeMonkeyLeet do you know?

**Comment by CodeMonkeyLeet** [comment]
My understanding is that in SGX1, the only EPC pages that are allocated to the enclave are whatever pages EADD had put into it prior to EINIT. The size of the linear address range of the enclave based on the alignment requirements does not incur EPC pages. The Linux SGX driver makes it a bit difficult to figure out since it permits overcommitment of EPC pages, but you could probably verify that on Windows.

**Comment by letmaik** [comment]
OK, so the working assumption is that in SGX the reported enclave size corresponds to allocated memory. What about simulation mode? The code path seems to be different there: https://github.com/openenclave/openenclave/blob/2490b0268f1dcf346dd977c81ac90ba73546ff9c/host/sgx/sgxload.c#L183-L196

**Comment by CodeMonkeyLeet** [comment]
> OK, so the working assumption is that in SGX the reported enclave size corresponds to allocated memory.

For SGX, the reported OE `enclave_t` size should be the size of the linear address range of the enclave, not the size of the total allocated pages prior to EINIT. I don't believe we track the latter anywhere.

For simulation mode, yes, it just mmaps the entire linear address range of the simulated enclave region on Linux; on Windows it commits the entire linear range, even though it could just reserve the range and commit only the pages being added to the enclave. That would give it semantics closer to the SGX implementation, but I don't know if there is an equivalent in Linux?

Essentially, in both the simulation and SGX modes, the enclave size is taken out of the virtual address range available to the process - in Windows terms, the memory range is _reserved_ - and only the pages added to the enclave need to be _committed_ (backed by physical memory pages). mmap doesn't have equivalent semantics though, and I don't know what the implications of mmapping a large address range for an anonymous mapping is (i.e. does it incur the cost of physical page mappings up front?) 

@jhand2, would you know?

**Comment by jhand2** [comment]
mmap uses demand paging, there is no physical memory backing until that page is accessed. But like you said, I don't believe OE tracks how much memory is actually in use, it just tracks the size of the address space.

**Comment by letmaik** [comment]
OK thanks you two, I'm going to do some more experiments to verify our assumptions. In the past I'm pretty sure I had issues with simulation mode where it would go OOM where with SGX it was fine, on the same VM.

**Comment by letmaik** [comment]
I did some tests on a DC8 with 29G available memory.

`NumHeapPages = 7340032` (28 GiB):
- SGX: runs fine, virtual memory of SGX-LKL is shown as 32.115 GiB.
- Simulation: fails immediately with `mmap failed mmap_size=68719476736 mflags=0x21 [../../host/sgx/sgxload.c:_allocate_enclave_memory:202]`.

`NumHeapPages = 4456448` (17 GiB):
- SGX: runs fine, virtual memory of SGX-LKL is shown as 32.115 GiB.
- Simulation: fails immediately with `mmap failed mmap_size=68719476736 mflags=0x21 [../../host/sgx/sgxload.c:_allocate_enclave_memory:202]`.

Same machine, but `NumHeapPages = 3932160` (15 GiB):
- SGX: runs fine, virtual memory of SGX-LKL is shown as 16.115 GiB.
- Simulation: fails immediately with `mmap failed mmap_size=34359738368 mflags=0x21 [../../host/sgx/sgxload.c:_allocate_enclave_memory:202]`.

Same machine, but `NumHeapPages = 2097152` (8 GiB):
- SGX: runs fine, virtual memory of SGX-LKL is shown as 16.115 GiB.
- Simulation: fails immediately with `mmap failed mmap_size=34359738368 mflags=0x21 [../../host/sgx/sgxload.c:_allocate_enclave_memory:202]`.

Same machine, but `NumHeapPages = 3932160` (7 GiB):
- SGX: runs fine, virtual memory of SGX-LKL is shown as 8.115 GiB.
- Simulation: runs fine, virtual memory of SGX-LKL is shown as 8.115 GiB.

In summary, SGX seems fine and as expected. However, simulation mode seems quite off. Essentially the maximum heap size I can test on DC8 in simulation mode is ~7.8G (leaving 200 MiB for the ELF image etc. to prevent jumping into the next power-of-two bucket). Is this a bug in OE?

**Comment by jhand2** [comment]
Can you get the following values from your VM?

`$ cat /proc/sys/vm/overcommit_memory`
`$ cat /proc/sys/vm/overcommit_ratio`
`$ cat /proc/meminfo | grep CommitLimit`
`$ cat /proc/meminfo | grep Committed_AS`

There could also be some nuance in the flags we pass to mmap, I'd have to look more into that.

**Comment by jhand2** [comment]
Basically if the value of `/proc/sys/vm/overcommit_memory` is 1, mmap will always overcommit. But if it's 0, the kernel gets to choose

(if the value is 2 it will never overcommit)

It looks like we can pass `MAP_NORESERVE` to mmap to get it to always overcommit for this specific allocation as long as the policy values isn't 2.

**Comment by letmaik** [comment]
```
$cat /proc/sys/vm/overcommit_memory
0
$cat /proc/sys/vm/overcommit_ratio
50
$cat /proc/meminfo | grep CommitLimit
CommitLimit:    16443608 kB
$cat /proc/meminfo | grep Committed_AS
Committed_AS:    1775028 kB
```

**Comment by jhand2** [comment]
Ok, so an overcommit_memory value of 0 means that the kernel uses a heuristic-based approach to decide whether to overcommit and how much. Maybe it has some limits on how much it will let a single process overcommit or something like that, I don't really know the in-kernel details of how it decides.

If you want, you can change OE's simulation mode loader to add the flag MAP_NORESERVE to the mmap call to test it out.

In host/sgx/sgxload.c around line 190 is the following:
```
        /* If no file descriptor, then perform anonymous mapping and double
         * the allocation size, so that BASE can be aligned on the SIZE
         * boundary. This isn't neccessary on hardware backed enclaves, since
         * the driver will do the alignment. */
        if (fd == -1)
        {
            mflags |= MAP_ANONYMOUS;
            if (oe_safe_mul_u64(mmap_size, 2, &mmap_size) != OE_OK)
            {
                OE_TRACE_ERROR(
                    "oe_safe_mul_u64 failed mmap_size = %ld", mmap_size);
                goto done;
            }
        }
```

You would just need to change the mflags line to `mflags |= MAP_ANONYMOUS | MAP_NORESERVE;`

If not, I can test it out but I probbly won't have time for a few days.

**Comment by letmaik** [comment]
I tried the flag, it works! Should I submit a PR to OE?

**Comment by jhand2** [comment]
> I tried the flag, it works! Should I submit a PR to OE?

That would be great, thanks! You can go ahead and submit the PR against master.

**Comment by letmaik** [comment]
Closing this issue and fixing upstream: https://github.com/openenclave/openenclave/pull/3041
